### PR TITLE
Update LTSM tutorial with proper description.

### PR DESCRIPTION
**Pull request by carriepl**
This adds a description about LSTMs in general as well as the specific model that is implemented in the LSTM tutorial.

There are two points still left address
-Validate which paper added the forget gate to the LSTM model and document it
-Validate with Cho about whether or not the input, forget and output gates should rely on the states on the memory cell. He said it is usually the case for the output gate but said nothing about the other gates. In Alex Graves' paper, it seems like all the gate activations usually rely on the state of the memory cell.

### CODE REVIEW

File: doc/lstm.txt [hunk 1]
```diff
@@ -1,14 +1,177 @@
...
+
+As previously mentionned, the provided scripts are used to train a LSTM
+recurrent neural on the Large Movie Review Dataset dataset.
+
+While the dataset is public, in this tutorial we provide a copy of the dataset
+that has previously been preprocessed according to the needs of this LSTM
+implementation. You can download this preprocessed version of the dataset
+using the script `download.sh
+<https://raw.githubusercontent.com/lisa-lab/DeepLearningTutorials/master/data/download.sh>`_
+and uncompress it.
```
Comment (nouiz) [inline doc/lstm.txt]: The script uncompress it. So you do not need the last part.

File: doc/lstm.txt [hunk 2]
```diff
@@ -1,14 +1,177 @@
...
+*vanishing gradients* where the gradient signal gets so small that learning
+either becomes very slow or stops working altogether. It can also make more
+difficult the task of learning long-term dependencies in the data.
+Conversely, if the weights in this matrix are large, it can lead to a
+situation where the gradient signal is so large that it can cause learning to
+diverge. This is often referred to as *exploding gradients*.
+
+These issues are the main motivation behind the LSTM model which introduces a
+new structure called a *memory cell* (see Figure 1 below). A memory cell is
+composed of four main elements: an input gate, a neuron with a self-recurrent
```
Comment (nouiz) [inline doc/lstm.txt]: I think there is more then 1 neurone. "a neuron" -> "neurons"?

Comment (carriepl) [inline doc/lstm.txt]: I do not think so. However, you can have a "memory cell block" (see http://web.eecs.utk.edu/~itamar/courses/ECE-692/Bobby_paper1.pdf page 12 paragraph 4.4). This is a group of memory cells that share both input gave and output gate.

**Comment by carriepl** [comment]
Updated the PR according to feedback from @nouiz

**Comment by carriepl** [comment]
@nouiz Can you give it one last look?

File: doc/lstm.txt [hunk 3]
```diff
@@ -22,23 +184,33 @@ The LSTM implementation can be found in the two following files :
 
 * `imdb.py <http://deeplearning.net/tutorial/code/imdb.py>`_ : Secondary script. Handles the loading and preprocessing of the IMDB dataset.
 
-Data
-====
+After downloading both the scripts, downloading the data and
+putting all those files in the same folder, the user can run the code by
```
Comment (nouiz) [inline doc/lstm.txt]: Do not tell downloading the data.

You can add after the execution command, that it will automatically download the data.

**Comment by carriepl** [comment]
Removed references to downloading the data

File: doc/lstm.txt [hunk 4]
```diff
@@ -1,14 +1,175 @@
...
+<http://ai.stanford.edu/~amaas/data/sentiment/>`_, sometimes known as the
+IMDB dataset.
+
+In this task, given a movie review, the model attempts to predict whether it
+is positive or negative. This is a binary classification task.
+
+Data
+++++
+
+As previously mentionned, the provided scripts are used to train a LSTM
```
Comment (kyunghyuncho) [inline doc/lstm.txt]: "mentionned" => "mentioned"

File: doc/lstm.txt [hunk 5]
```diff
@@ -1,14 +1,175 @@
...
+recurrent neural on the Large Movie Review Dataset dataset.
```
Comment (kyunghyuncho) [inline doc/lstm.txt]: "recurrent neural" => "recurrent neural network"

File: doc/lstm.txt [hunk 6]
```diff
@@ -1,14 +1,175 @@
...
+====
+
+In a *traditional* recurrent neural network, during the gradient
+back-propagation phase, the gradient signal can end up being multiplied a
+large number of times (as many as the number of timesteps) by the weight
+matrix associated with the connections between the neurons of the recurrent
+hidden layer. This means that, the magnitude of weights in the transition
+matrix can have a strong impact on the learning process.
+
+If the weights in this matrix are small, it can lead to a situation called
```
Comment (kyunghyuncho) [inline doc/lstm.txt]: It's probably better to explicitly say that "the leading eigenvalue of the weight matrix is small (< 1)" or "large (> 1)".

File: doc/lstm.txt [hunk 7]
```diff
@@ -1,14 +1,175 @@
...
+diverge. This is often referred to as *exploding gradients*.
+
+These issues are the main motivation behind the LSTM model which introduces a
+new structure called a *memory cell* (see Figure 1 below). A memory cell is
+composed of four main elements: an input gate, a neuron with a self-recurrent
+connection (a connection to itself), a forget gate and an output gate. The
+self-recurrent connection has a weight of 1.0 and ensures that, barring any
+outside interference, the state of a memory cell can remain constant from one
+timestep to another. The gates serve to modulate the interactions between the
+memory cell and itself and its environment. The input gate can allow incoming
```
Comment (kyunghyuncho) [inline doc/lstm.txt]: "memory cell and itself" => "memory cell itself"?

**Comment by kyunghyuncho** [comment]
@carriepl, there were some typos in the pull request. Once they are fixed, I think we can merge it right away.

**Comment by carriepl** [comment]
Fixed the typos

**Pull request merged**